### PR TITLE
Fix VPN redirects (Fixes #13460)

### LINKED ISSUE #13460: VPN RC Redirect changes
Certain pages don't maintain the correct locale from the old `products/vpn/more/*` to the vpn resource center. I am assuming this is happening because certain locales haven't been fully translated for the resource center.

Adria created this [spreadsheet](https://docs.google.com/spreadsheets/d/1Ud40NM11LkxEs2dm0RAMqiRQguLLPc7n-IBtzqmOb-A/edit?usp=sharing) lists out which pages don't have equivalent translated pages.

Examples:
`https://www.mozilla.org/ar/products/vpn/more/vpn-or-proxy/` redirects to `https://www.mozilla.org/en-US/products/vpn/resource-center/the-difference-between-a-vpn-and-a-web-proxy/`

### PR DESCRIPTION
**Pull request by nathan-barrett**
## One-line summary
Re-adds pages at `vpn/more/*` which were removed in #12938

## Significant changes and points to review
- Adds a view in `products/views.py` that takes three arguments (template, resource-center slug, and ftl_files) - this way, when navigating to a `vpn/more/*` url - this view checks the associated page in the VPN resource center and checks if the active locale is available. If it is available, the user is redirected to the resource center, if not, the user is sent to the correct local with the `vpn/more/*` url.

## Issue / Bugzilla link
#13460

## Testing
If the locale is available in smartling/contentful the page will be redirected to the corresponding article in the VPN resource center, otherwise it will navigate to the first page.

Test different locales to make sure both situations work. Some locales that `do not` trigger the redirect are: `en-GB, ar and cs`

- http://localhost:8000/en-US/products/vpn/more/what-is-an-ip-address/ -> /products/vpn/resource-center/what-is-an-ip-address/
- http://localhost:8000/en-US/products/vpn/more/what-is-a-vpn/  -> /products/vpn/resource-center/what-is-a-vpn/
- http://localhost:8000/en-US/products/vpn/more/vpn-or-proxy/ -> /products/vpn/resource-center/the-difference-between-a-vpn-and-a-web-proxy/
- http://localhost:8000/en-US/products/vpn/more/when-to-use-a-vpn/ -> /products/vpn/resource-center/

### CODE REVIEW

File: bedrock/products/urls.py [hunk 1]
```diff
@@ -25,6 +25,51 @@
     path("vpn/mobile/ios/", views.VPNIosView.as_view(), name="products.vpn.platforms.ios"),
     path("vpn/mobile/android/", views.VPNAndroidView.as_view(), name="products.vpn.platforms.android"),
     page("vpn/ipad/", "products/vpn/platforms/ipad.html", ftl_files=["products/vpn/platforms/ipad", "products/vpn/shared"]),
+    # Evergreen SEO articles (issue #10224)
+    path(
```
Comment (pmac) [inline bedrock/products/urls.py:29]: You could compress this into one URL conf and a mapping of slug to old template and ftl files in the views.py file. 

```python
path("vpn/more/<slug:slug>/", views.vpn_resource_center_redirect, name="products.vpn.more.redirect")
```
Then in views.py
```python
VPNRC_SLUGS = {
    "what-is-an-ip-address": {
        "slug": "what-is-an-ip-address",
        "old_template": "products/vpn/more/ip-address.html",
        "ftl_files": ["products/vpn/more/ip-address", "products/vpn/shared"],
    },
    "vpn-or-proxy": {
        "slug": "the-difference-between-a-vpn-and-a-web-proxy",
        "old_template": "products/vpn/more/vpn-or-proxy.html",
        "ftl_files": ["products/vpn/more/vpn-or-proxy", "products/vpn/shared"],
    },
}
```
You've already done the work here, so you don't really need to do this, but for future reference that's how I'd probably do it. This works just fine though.

File: bedrock/products/views.py [hunk 2]
```diff
@@ -516,3 +526,14 @@ def monitor_landing_page(request):
...
     return l10n_utils.render(request, template_name, context, ftl_files=ftl_files)
+
+
+@require_safe
+def vpn_resource_center_redirect(request, slug, old_template, ftl_files):
+    locale = l10n_utils.get_locale(request)
+    slug = "what-is-an-ip-address"
+    redirect_link = f"/products/vpn/resource-center/{slug}/"
+    if active_locale_available(slug, locale):
+        return redirect(redirect_link)
```
Comment (pmac) [inline bedrock/products/views.py:537]: we know the requested locale, might as well send them directly to it.

File: bedrock/products/urls.py (hunk 1 above)
Comment (nathan-barrett) [reply bedrock/products/urls.py:29]: This is great, thanks!

File: bedrock/products/views.py [hunk 3]
```diff
@@ -516,3 +526,14 @@ def monitor_landing_page(request):
     context = {"breach_scan_url": breach_scan_url}
 
     return l10n_utils.render(request, template_name, context, ftl_files=ftl_files)
+
+
+@require_safe
+def vpn_resource_center_redirect(request, slug, old_template, ftl_files):
+    locale = l10n_utils.get_locale(request)
+    slug = "what-is-an-ip-address"
+    redirect_link = f"/products/vpn/resource-center/{slug}/"
```
Comment (pmac) [inline bedrock/products/views.py:535]:
```suggestion
    redirect_link = f"/{locale}/products/vpn/resource-center/{slug}/"
```

**Review by stephaniehobson — APPROVED** [review]
R+

File: bedrock/products/views.py [hunk 4]
```diff
@@ -516,3 +526,14 @@ def monitor_landing_page(request):
     context = {"breach_scan_url": breach_scan_url}
 
     return l10n_utils.render(request, template_name, context, ftl_files=ftl_files)
+
+
+@require_safe
+def vpn_resource_center_redirect(request, slug, old_template, ftl_files):
+    locale = l10n_utils.get_locale(request)
+    slug = "what-is-an-ip-address"
```
Comment (stephaniehobson) [inline bedrock/products/views.py:534]: The slug is hard coded here so everything is forwarding to it >_< 
```suggestion
```

File: bedrock/products/templates/products/vpn/more/ip-address.html [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+<p>{{ ftl('vpn-ip-address-when-the-internet', sr='https://www.siliconrepublic.com/comms/ip-addresses-running-out',
+  variety='https://variety.com/2019/digital/news/u-s-households-have-an-average-of-11-connected-devices-and-5g-should-push-that-even-higher-1203431225/')
+  }}</p>
+<p>{{ ftl('vpn-ip-address-we-now-have') }}</p>
+
+<h3>{{ ftl('vpn-ip-address-should-you-hide') }}</h3>
+<p>{{ ftl('vpn-ip-address-you-dont-need',
+  congress='https://blog.mozilla.org/en/mozilla/internet-policy/u-s-broadband-privacy-rules-will-fight-protect-user-privacy/',
+  doh='https://support.mozilla.org/kb/firefox-dns-over-https', firefox=url('firefox.new')) }}</p>
+<p>{{ ftl('vpn-ip-address-there-are-also', url=url('products.vpn.more.when-to-use')) }}</p>
```
Comment (stephaniehobson) [inline bedrock/products/templates/products/vpn/more/ip-address.html:66]: This url value fails now 😭

File: bedrock/products/views.py [hunk 6]
```diff
@@ -516,3 +526,14 @@ def monitor_landing_page(request):
...
+    redirect_link = f"/{locale}/products/vpn/resource-center/{slug}/"
+    if active_locale_available(slug, locale):
+        return redirect(redirect_link)
+    else:
+        return l10n_utils.render(request, template=old_template, ftl_files=ftl_files)
```
Comment (stephaniehobson) [inline bedrock/products/views.py:539]: In cases where neither template is translated we are currently displaying the old template at the old URL in English at their locale. I think in these cases we usually forward them to the canonical English locale. 

Example: http://localhost:8000/ach/products/vpn/more/what-is-a-vpn/

Comment (stephaniehobson) [reply bedrock/products/views.py:539]: I don't think this is a result of the changes Pmac and I merged as reverting it didn't fix it locally but I didn't spend too long looking into it.

**Review by stephaniehobson — CHANGES_REQUESTED** [review]
All pages are forwarding to the same slug. Looks like maybe a line of debugging code was left?

Once that's fixed there's an error displaying [this old template](http://localhost:8000/en-CA/products/vpn/more/what-is-an-ip-address/ ). 

I'm also not sure the behaviour when there is [no match found](http://localhost:8000/ach/products/vpn/more/what-is-a-vpn/) is quite what we want.

This is a real brain bender. No surprise it's complicated!

**Comment by pmac** [comment]
FYI: your commit message mentions https://github.com/mozilla/bedrock/issues/13640 but the right one is https://github.com/mozilla/bedrock/issues/13460

**Comment by nathan-barrett** [comment]
Thanks for reviews and help with this, @pmac and @stephaniehobson! Will work and fix this today

**Comment by nathan-barrett** [comment]
Updated and added some tests - is good to re-review - thanks @stephaniehobson

**Comment by stephaniehobson** [comment]
There's a merge conflict with static-bundles ☹️

**Comment by nathan-barrett** [comment]
@stephaniehobson - rebased and re-pushed. Hopefully that fixes any conflicts

File: bedrock/products/views.py [hunk 7]
```diff
@@ -532,3 +542,39 @@ def mozsocial_waitlist_page(request):
...
+            "slug": "5-reasons-you-should-use-a-vpn",
+            "old_template": "products/vpn/more/when-to-use.html",
+            "ftl_files": ["products/vpn/more/when-to-use-a-vpn", "products/vpn/shared"],
+        },
+    }
+    locale = l10n_utils.get_locale(request)
+    curr_page = VPNRC_SLUGS[slug]
+    rc_slug = curr_page["slug"]
+    redirect_link = f"{locale}/products/vpn/resource-center/{rc_slug}/"
+    if active_locale_available(rc_slug, locale):
```
Comment (stephaniehobson) [inline bedrock/products/views.py:575]: I think we need to think through what the desired behaviour here is so we can check that the logic is matching up.

I'll call the new templates rc and the old more

When a /resource-centre url is requested the user should have the rc article displayed in the requested language. 
- If the rc article is not available in their requested language bedrock should display the more article is available in their language. 
- If neither is available in their language bedrock should forward to the English rc article.

When a /more url is requested the user should be forwarded to the /resource-centre url. The above logic then applies.

Is that right?

**Review by stephaniehobson — CHANGES_REQUESTED** [review]
The redirects are now including both the old and new paths in the URL. So `http://localhost:8000/en-US/products/vpn/more/what-is-an-ip-address/`  is redirecting to `http://localhost:8000/en-US/products/vpn/more/what-is-an-ip-address/en-US/products/vpn/resource-center/what-is-an-ip-address/`

File: bedrock/products/views.py (hunk 7 above)
Comment (nathan-barrett) [reply bedrock/products/views.py:575]: Yes, that is right. I'll add a comment to the view to make that clear

Comment (nathan-barrett) [reply bedrock/products/views.py:575]: I think the focus of this PR was to just handle the redirects for the /more/ urls - but I can check with adria to make sure.

**Comment by nathan-barrett** [comment]
OK, fixed errors in tests and checked in with Adria that the links are working as expected - ready for re-review @stephaniehobson

File: bedrock/products/views.py (hunk 7 above)
Comment (stephaniehobson) [reply bedrock/products/views.py:596]: Re-reading the "significant changes" section of the PR I can see that what I outlined is not the behaviour you outlined and it is desired to display something at the /more/ URL in some cases. I'm sorry I made a bad logical leap about the desired behaviour.

**Review by stephaniehobson — CHANGES_REQUESTED** [review]
Okay I think we have 3 use cases here when a /more/ URL is requested:

1. _more_ articles with a _rc_ equivalent in the same language [fr](http://localhost:8000/fr/products/vpn/more/what-is-an-ip-address/)
2.  _more_ articles without a _rc_ equivalent in the same language [ar](http://localhost:8000/ar/products/vpn/more/what-is-an-ip-address/)
3.  _more_ articles that do not exist in the requested language [ach](http://localhost:8000/ach/products/vpn/more/what-is-an-ip-address/)

---

1. 💯 Is working as expected
2. 💯 I made a bad logical leap about how this was supposed to work and it turns out it is also working as expected 🤦‍♀️ 
3. At the moment these are displaying the en-US content at the requested locale. In cases where we do not have the requested language we should forward to the en-US content. For example: https://www.mozilla.org/ach/about/manifesto/ forwards to https://www.mozilla.org/en-US/about/manifesto/ 

Just one edge case to smooth off and we're good to go.

**Comment by nathan-barrett** [comment]
Ok, the changes I made should fix that edge case @stephaniehobson - when you test locally turn `Dev=False` and it should solve the probelm

**Review by stephaniehobson — APPROVED** [review]
R+ ↪️
